src/RamTransforms.cpp

```diff
@@ -561,183 +561,140 @@ bool HoistAggregateTransformer::hoistAggregate(RamProgram& program) {
     // flag to determine whether the RAM program has changed
     bool changed = false;
 
-    // hoist aggregate if doesn't depend on anything
+    // hoist a single aggregate to an outer scope
+    // most outer permissible aggregate is hoisted
     visitDepthFirst(program, [&](const RamQuery& query) {
-        int currLevel = -1;
-        int oldLevel = -1;
-        bool hoist = false;
-        bool newIndex = false;
-
-        // Fields for new aggregate
+        // aggregate description
+        int aggLevel = -1;
         AggregateFunction newFun;
         const RamRelation* newRef;
         RamExpression* newExp;
         RamCondition* newCond;
-        std::vector<RamExpression*> newPattern;
+
+        // is indexed aggregate operation
+        bool isAggIndexed = false;
+        std::vector<std::unique_ptr<RamExpression>> newPattern;
+
+        // new level of aggregate
+        int newLevel = -1;
 
         // Tracking aggregates seen to determine whether to
         // hoist or not
         std::vector<int> aggIds;
 
-        // Removing an aggregate that can be hoisted if it exists
-        std::function<std::unique_ptr<RamNode>(std::unique_ptr<RamNode>)> aggRemover =
+        // check whether we have aggregates only to
+        // dependent loop
+        auto checkForAggregates = [&](int start, int end) {
+            bool allAggregates = true;
+            for (int i = start + 1; i <= end; i++) {
+                if (!(std::find(aggIds.begin(), aggIds.end(), i) != aggIds.end())) {
```

Comment:
This could be simplified
```suggestion
                if (std::find(aggIds.begin(), aggIds.end(), i) == aggIds.end()) {
```